File: 백준/Silver/12101. 1， 2， 3 더하기 2/1， 2， 3 더하기 2.py
"""
백트로 풀자

기저조건은 total이 n이면 스탑

1, 2, 3의 개수를 계속 들고 다니다가
기저조건 만나면 answer에 넣기

그리고 answer 정렬하고 맞는거 빼기
"""
import sys

input = lambda: sys.stdin.readline().strip()

n, k = map(int, input().split())

lst = [0, 1, 2, 4]

for i in range(4, 11):
    lst.append(lst[i-1] + lst[i-2] + lst[i-3])

answer = []


def recur(arr, total):
    if total > n:
        return
    if total == n:
        answer.append(arr)
        return
    recur(arr + [1], total + 1)
    recur(arr + [2], total + 2)
    recur(arr + [3], total + 3)


if k > lst[n]:
    print(-1)
else:
    recur([], 0)
    answer = sorted(answer)
    print('+'.join(map(str, answer[k-1])))

File: 백준/Silver/12101. 1， 2， 3 더하기 2/1， 2， 3 더하기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	private static int n;
	private static int k;
	private static ArrayList<ArrayList<Integer>> answer;
	private static int cnt;
	
	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		
		n = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		
		cnt = 0;
		answer = new ArrayList<ArrayList<Integer>>();
		ArrayList<Integer> lst = new ArrayList<>();
		recur(lst, 0);
		
		if (k > cnt) {
			System.out.println(-1);
		} else {
			String str = "";
			for (int i = 0; i < answer.get(k-1).size(); i++) {
				str = str.concat(answer.get(k-1).get(i).toString());
				str = str.concat("+");
			}
			
			System.out.println(str.substring(0, str.length()-1));
		}
	} // end of main

	private static void recur(ArrayList<Integer> arr, int total) {
		if (total > n)
			return;
		if (total == n) {
			ArrayList<Integer> newarr = new ArrayList<>(arr);
			answer.add(newarr);
			cnt++;
			return;
		}
		arr.add(1);
		recur(arr, total + 1);
		arr.remove(arr.size()-1);
		arr.add(2);
		recur(arr, total + 2);
		arr.remove(arr.size()-1);
		arr.add(3);
		recur(arr, total + 3);
		arr.remove(arr.size()-1);
	}
} // end of class